brushfire-spark/src/main/scala/com/stripe/brushfire/spark/Trainer.scala

```diff
@@ -0,0 +1,226 @@
+
+    val collectFeatures: Instance[K, V, T] => Iterable[(Bucket, splitter.S)] = { instance =>
+      val features = instance.features.mapValues(splitter.create(_, instance.target))
+      for {
+        (treeIndex, tree) <- treeMap.toList
+        repetition = sampler.timesInTrainingSet(instance.id, instance.timestamp, treeIndex)
+        i <- 1 to repetition
+        leaf <- tree.leafFor(instance.features).toList
+        if stopper.shouldSplit(leaf.target) && stopper.shouldSplitDistributed(leaf.target)
+        (feature, stats) <- features
+        if sampler.includeFeature(feature, treeIndex, leaf.index)
+      } yield {
+        (treeIndex, leaf.index, feature) -> stats
+      }
+    }
+
+    val split: (Bucket, splitter.S) => Iterable[(Int, Map[Int, ScoredSplit])] = { (bucket, stats) =>
+      val (treeIndex, leafIndex, feature) = bucket
+      for {
+        leaf <- treeMap(treeIndex).leafAt(leafIndex).toList
+        rawSplit <- splitter.split(leaf.target, stats)
+      } yield {
+        val (split, goodness) = evaluator.evaluate(rawSplit)
+        treeIndex -> Map(leafIndex -> (feature, split, goodness))
+      }
+    }
+
+    val emptySplits: RDD[(Int, Map[Int, ScoredSplit])] =
+      context.parallelize(Seq.tabulate(sampler.numTrees)(_ -> Map.empty[Int, ScoredSplit]))
+
+    val growTree: (Int, Map[Int, ScoredSplit]) => (Int, Tree[K, V, T]) = { (treeIndex, leafSplits) =>
+      val newTree =
+        treeMap(treeIndex)
+          .growByLeafIndex { index =>
+            for {
+              (feature, split, _) <- leafSplits.get(index).toList
+              (predicate, target) <- split.predicates
+            } yield {
+              (feature, predicate, target)
+            }
+          }
+      treeIndex -> newTree
+    }
+
+    // Ugh. We could also wrap splitter.S in some box...
+    implicit val existentialClassTag: ClassTag[splitter.S] =
+      scala.reflect.classTag[AnyRef].asInstanceOf[ClassTag[splitter.S]]
+
+    val newTrees = trainingData
+      .flatMap(collectFeatures)
+      .reduceByKey(splitter.semigroup.plus(_, _))
+      .flatMap(split.tupled)
+      .union(emptySplits)
+      .sumByKey
+      .map(growTree.tupled)
+      .cache()
+
+    copy(trees = newTrees)
+  }
+}
+
+object Trainer {
+  val MaxParallelism = 20
+
+  def apply[K: Ordering, V, T: Monoid: ClassTag](trainingData: RDD[Instance[K, V, T]], sampler: Sampler[K]): Trainer[K, V, T] = {
+    val sc = trainingData.context
+    val initialTrees = Vector.tabulate(sampler.numTrees) { _ -> Tree.empty[K, V, T](Monoid.zero) }
+    val parallelism = scala.math.min(MaxParallelism, sampler.numTrees)
+    Trainer(
+      trainingData,
+      sampler,
+      sc.parallelize(initialTrees, parallelism))
+  }
+
+  implicit class ExtraRDDOps[A](val rdd: RDD[A]) extends AnyVal {
+    def sumByKey[K, V](implicit ev: A <:< (K, V), ctK: ClassTag[K], ctV: ClassTag[V], V: Monoid[V]): RDD[(K, V)] =
+      rdd.asInstanceOf[RDD[(K, V)]].foldByKey(V.zero)(V.plus)
```

Comment:
Yes, it *should* be efficient. Specifically, the docs state that the "zero" may be added an arbitrary number of times to the result and cannot change it. So, it isn't `fold` in the Scala sense where the first parameter is an initial value, but rather it must be a zero in the sense of a monoid. It's kind of an unfortunate name though, since it implies inefficiency :\